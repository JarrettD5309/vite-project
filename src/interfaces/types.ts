interface BookListItem {
  title: string;
  url: string;
  author: string;
  num_comments: number;
  points: number;
  objectID: number;
}

interface ItemProps {
  item: BookListItem;
  onRemoveItem: (item: BookListItem) => void;
}

interface ListProps {
  list: BookListItem[];
  onRemoveItem: (item: BookListItem) => void;
  onSortAscendDescend: (col: ColName) => void;
}

interface ListTitleProps {
  onSortAscendDescend: (col: ColName) => void;
}

interface InputWithLabelProps {
  id: string;
  value: string;
  isFocused: boolean;
  type?: string;
  onInputChange: (e: React.ChangeEvent<HTMLInputElement>) => void
}

interface LastSearchProps {
  pastSearchArr: string[];
  setSearchTerm: (newVal: string) => void;
  handlePreviousSearch: (pastTerm: string) => void;
}

interface SearchFormProps {
  searchTerm: string;
  onSearchInput: (e: React.ChangeEvent<HTMLInputElement>) => void;
  onSearchSubmit: (e: React.FormEvent<HTMLFormElement>) => void;
}

interface StoryReducerActionObj {
  type: StoryAction;
  payload?: BookListItem[]
}

interface StoryReducerObj {
  data: BookListItem[];
  isLoading: boolean;
  isError: boolean;
}

interface StyledColumnProps {
  width: string;
}

interface WelcomeObj {
  greeting: string;
  title: string;
}

enum ColName {
  TITLE = 'title',
  AUTHOR = 'author',
  COMMENTS = 'num_comments',
  POINTS = 'points'
}

enum StoryAction {
  REMOVE_STORY = 'REMOVE_STORY',
  SORT_ASCEND_DESCEND = 'SORT_ASCEND_DESCEND',
  STORIES_FETCH_INIT = 'STORIES_FETCH_INIT',
  STORIES_FETCH_FAILURE = 'STORIES_FETCH_FAILURE',
  STORIES_FETCH_MORE_SUCCESS = 'STORIES_FETCH_MORE_SUCCESS',
  STORIES_FETCH_SUCCESS = 'STORIES_FETCH_SUCCESS'
}

export {
  type BookListItem,
  type ItemProps,
  type ListProps,
  type ListTitleProps,
  type InputWithLabelProps,
  type LastSearchProps,
  type SearchFormProps,
  type StoryReducerActionObj,
  type StoryReducerObj,
  type StyledColumnProps,
  type WelcomeObj,
  StoryAction,
  ColName
};