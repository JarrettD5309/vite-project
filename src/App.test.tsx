import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';

import { BookListItem, StoryAction, StoryReducerActionObj, StoryReducerObj } from './interfaces/types.ts';

import {
    storiesReducer,
    Item,
} from './App.tsx';

const storyOne: BookListItem = {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0
};

const storyTwo: BookListItem = {
    title: 'Redux',
    url: 'https://reduxjs.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1
};

const stories: BookListItem[] = [storyOne, storyTwo];

describe('something truthy and falsy', () => {
    it('true to be true', () => {
        expect(true).toBeTruthy();
    });

    it('false to be false', () => {
        expect(false).toBeFalsy();
    });
});

describe('storiesReducer', () => {
    it('sets isLoading to TRUE and isError to FALSE on FETCH INIT', () => {
        const storyObj2: StoryReducerObj = {
            isError: false,
            isLoading: false,
            data: []
        };

        const storyActionObj2: StoryReducerActionObj = {
            type: StoryAction.STORIES_FETCH_INIT
        };

        const newState: StoryReducerObj = storiesReducer(storyObj2, storyActionObj2);
        
        const expectedState: StoryReducerObj = {
            isLoading: true,
            isError: false,
            data: []
        };

        expect(newState).toStrictEqual(expectedState);
    });

    it('sets data to payload, isLoading to FALSE, isError to FALSE on FETCH SUCCESS', () => {
        const storyObj3: StoryReducerObj = {
            data: [],
            isError: true,
            isLoading: true
        };

        const storyActionObj3: StoryReducerActionObj = {
            type: StoryAction.STORIES_FETCH_SUCCESS,
            payload: stories
        };

        const newState: StoryReducerObj = storiesReducer(storyObj3, storyActionObj3);

        const expectedState: StoryReducerObj = {
            data: stories,
            isError: false,
            isLoading: false
        };

        expect(newState).toStrictEqual(expectedState);
    });

    it('sets isLoading to FALSE and isError to TRUE on FAILURE', () => {
        const storyObj4: StoryReducerObj = {
            data: [],
            isError: false,
            isLoading: true
        };

        const storyActionObj4: StoryReducerActionObj = {type: StoryAction.STORIES_FETCH_FAILURE};

        const newState: StoryReducerObj = storiesReducer(storyObj4, storyActionObj4);

        const expectedState: StoryReducerObj = {
            ...storyObj4,
            isLoading: false,
            isError: true
        };

        expect(newState).toStrictEqual(expectedState);
    });

    it('removes a story from all stories on REMOVE STORY', () => {

        const storyObj1: StoryReducerObj = {
            isError: false,
            isLoading: false,
            data: stories
        };

        const storyActionObj1: StoryReducerActionObj = {
            type: StoryAction.REMOVE_STORY,
            payload: [storyOne]
        };

        const newState: StoryReducerObj = storiesReducer(storyObj1, storyActionObj1);

        const expectedState: StoryReducerObj = {
            ...storyObj1,
            data: [storyTwo]
        };

        expect(newState).toStrictEqual(expectedState);

    });
});

describe('Item', () => {
    it('renders all properties', () => {
        render(<Item item={storyOne} onRemoveItem={(x) => x}/>);

        expect(screen.getByText('Jordan Walke')).toBeInTheDocument();
        expect(screen.getByText('React')).toHaveAttribute(
            'href',
            'https://reactjs.org/'
        );
    });

    it('renders a clickable dismiss button', () => {
        render(<Item item={storyOne} onRemoveItem={(x) => x}/>);

        expect(screen.getByRole('button')).toBeInTheDocument();
    });
});