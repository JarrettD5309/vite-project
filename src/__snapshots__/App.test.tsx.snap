// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > renders snapshot 1`] = `
<div
  class="sc-eCstZk eBcclB"
>
  <h1
    class="sc-jSguLX bSCAos"
  >
    Hey
     
    React
  </h1>
  <h1
    class="sc-jSguLX bSCAos"
  >
    Hello World
  </h1>
  <h1
    class="sc-jSguLX bSCAos"
  >
    My Hacker Stories with 
    0
     comments.
  </h1>
  <form
    class="sc-pGaPU jXbBNc"
  >
    <label
      class="sc-fubDmA OksbU"
      for="search"
    >
      Search:
    </label>
     
    <input
      class="sc-gKsecS hfckNu"
      id="search"
      type="text"
      value="React"
    />
    <button
      class="sc-bdfCDU sc-dlfmHC fgeRLE dZwrtt"
      type="submit"
    >
      Submit
    </button>
  </form>
  <p>
    Loading...
  </p>
</div>
`;

exports[`SearchForm > renders snapshot 1`] = `
<form
  class="sc-pGaPU jXbBNc"
>
  <label
    class="sc-fubDmA OksbU"
    for="search"
  >
    Search:
  </label>
   
  <input
    class="sc-gKsecS hfckNu"
    id="search"
    type="text"
    value="React"
  />
  <button
    class="sc-bdfCDU sc-dlfmHC fgeRLE dZwrtt"
    type="submit"
  >
    Submit
  </button>
</form>
`;
